### PR TITLE
Issue 62

### PR DESCRIPTION
**Pull request by jofeinstein**
Added features to allow users, managers, and PIs to remove users from a project. 

Requests:
- Users can choose to request a removal from a project by clicking "Leave Project" in a project detail view
        - PIs cannot request to leave a project
        - Managers may request to leave if there are more than one manager in a project
- Managers and PIs can remove users from the "Remove Users" Page
- Remove users page now shows users with pending removal requests
- A user can only have a single active project removal request
- After a project removal request has been completed, the user will still appear in the allocation detail page until the users AllocationUserStatusChoice is set to something other than 'Active'
- Created a new table showing removed users' usages called "Users in Allocation and Removed from Project" in the allocation detail page
- Emails are sent to the user/PI/manager/admin when removal requests are made
- Requests are created using new ProjectRemovalRequestRunner in `coldfront/core/project/utils.py`

Approvals:
- added new tab for admins and staff called "Project Removal Requests"
     - staff only have read access
- can view both pending and completed requests on different pages
- updating the status of the request is a 2 part process
		- requests start with status "Pending"
		- admins must first click "Update" to change the status from "Pending" to "Processing"
		- admins must then perform the necessary actions on the cluster to remove the user from the project
		- once finished, admins click update again and change the status from "Processing" to "Complete"
			- There are two new views to facilitate this "ProjectRemovalRequestUpdateStatusView" and "ProjectRemovalRequestCompleteStatusView"
- emails are sent to the user/PI/manager when the request is marked as "Complete"

Database Changes:
- created two new models in project/models.py to keep track of removal requests
- ProjectUserRemovalRequest and ProjectUserRemovalRequestStatusChoice

Commands:
`python manage.py create_project_removal_request_statuses`
	- create 3 ProjectUserRemovalRequestStatusChoice with names Pending, Processing, Complete.
`python manage.py create_staff_group`
	- updates staff group to include project.view_projectuserremovalrequest permission

Tests:
- added tests for the ProjectRemovalRequestRunner and various new and updated views
- two test files in `coldfront/core/project/tests/ `
     - test_project_removal_request_runner.py and test_project_removal_request_views.py

Miscellaneous:
	- users with active removal requests cannot be added back to the same project
	- allocation usage remains the same after a user has been removed
		- under allocation detail view, can see usages of removed users
	- Users with pending removal requests can still view project detail page
	        - project still appears on home page and ProjectListView

### CODE REVIEW

File: coldfront/core/project/tests/test_project_removal_request_views.py [hunk 1]
```diff
@@ -0,0 +1,611 @@
+from django.test import TestCase
```
Comment (matthew-li) [inline coldfront/core/project/tests/test_project_removal_request_views.py:1]: Can you split these into one class per view and move this file under `tests.test_views`? The overlapping `setUp` logic can be extracted into a base class that they each inherit from.

**Review by matthew-li — CHANGES_REQUESTED** [review]
Thanks for getting this in, and especially for the thorough tests! It looks good overall, but I've left a range of comments throughout, mostly minor/organizational.

I'll test it manually soon.

File: coldfront/core/project/views.py [hunk 2]
```diff
@@ -3754,3 +3821,299 @@ def get(self, request, *args, **kwargs):
...
+            f'Project {self.project_removal_request_obj.project_user.project.name} '
+            f'has been marked as {status}.')
+        messages.success(self.request, message)
+
+        if EMAIL_ENABLED and status == 'Complete':
+            manager_pi_queryset = project_obj.projectuser_set.filter(
+                role__name__in=['Manager', 'Principal Investigator'],
+                status__name='Active')
+
+            for proj_user in list(chain(manager_pi_queryset, [removed_user])):
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/project/views.py:4095]: This way the greeting is different for each person, ie "Dear Joseph Feinstein" and "Dear Matthew Li"

File: coldfront/templates/common/navbar_nonadmin_staff.html [hunk 3]
```diff
@@ -17,6 +17,15 @@
     {% if perms.project.can_review_pending_project_reviews %}
     <a id="navbar-project-reviews" class="dropdown-item" href="{% url 'project-review-list' %}">Project Reviews</a>
     {% endif %}
+    {% if perms.allocation.can_review_allocation_requests %}
+    <a id="navbar-allocation-requests" class="dropdown-item" href="{% url 'allocation-request-list' %}">
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/templates/common/navbar_nonadmin_staff.html:21]: Yea I think that's a relic from branching off of an older develop branch

File: coldfront/core/project/forms.py [hunk 4]
```diff
@@ -1078,3 +1080,37 @@ def clean_final_name(self):
             raise forms.ValidationError(
                 f'A project with name {final_name} already exists.')
         return final_name
+
+
+class ProjectRemovalRequestSearchForm(forms.Form):
+    project_name = forms.CharField(label='Project Title',
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/project/forms.py]: Changed it to 'Project Name'

File: coldfront/core/project/templates/project/project_removal_request_complete_status.html [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    <div class="card mb-3">
+        <div class="card-header">
+            <h2>
+                <i class="fas fa-user-check" aria-hidden="true"></i>
+                Update
+            </h2>
+        </div>
+        <div class="card-body">
+            <p class="card-text text-justify">
+                Please change the status to "Complete" and type in the user's username to confirm.
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/project/templates/project/project_removal_request_complete_status.html]: Just an artifact from a previous iteration. A double confirmation seemed unnecessary.

File: coldfront/core/project/views.py [hunk 6]
```diff
@@ -1060,65 +1104,88 @@ def get(self, request, *args, **kwargs):
...
+        if project_obj.projectuser_set.filter(
+                user=self.request.user,
+                role__name='User',
+                status__name='Active').exists():
+            return True
+
+        if project_obj.projectuser_set.filter(
+                user=self.request.user,
+                role__name='Manager',
+                status__name='Active').exists() and len(project_obj.projectuser_set.filter(role__name='Manager')) > 1:
```
[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/project/views.py]: Not sure it makes sense to create a new method. I think the cases are slightly different in different places.

[the remark this answers is not in the document]
Comment (jofeinstein) [reply coldfront/core/project/views.py]: I'll remember the count method for the future

**Comment by jofeinstein** [comment]
@matthew-li I've gone through and made the requested changes. I also finished working on the 'project-remove-users' page.

File: coldfront/core/project/templates/project/project_remove_users.html [hunk 7]
```diff
@@ -1,73 +1,188 @@
...
-{% endif %}
-<script>
-$("#selectAll").click(function() {
-  $("input[name^='userform-']").prop('checked', $(this).prop('checked'));
-});
+        <p></p>
+        <td>
+            Below is a list of users eligible to be removed from project {{ project.title }}. To initiate a
+            removal request, click the "Remove" button next to the user to remove and click
+            "Submit User Removal Request" in the confirmation window. Once a removal
```
Comment (matthew-li) [inline coldfront/core/project/templates/project/project_remove_users.html:62]: "Submit" might be sufficient for the button.

File: coldfront/core/project/views.py [hunk 8]
```diff
@@ -1055,93 +1092,147 @@ def dispatch(self, request, *args, **kwargs):
...
+
+            if runner_result:
+                request_runner.send_emails()
+                for message in success_messages:
+                    messages.success(request, message)
+            else:
+                for message in error_messages:
+                    messages.error(request, message)
+        except Exception as e:
+            self.logger.exception(e)
```
Comment (matthew-li) [inline coldfront/core/project/views.py:1230]: This has caused problems for me in the past. This class has no attribute `logger`.

It's safer to define `logger` at the top of the file, and then reference it without this concern.

File: coldfront/core/project/views.py [hunk 9]
```diff
@@ -887,6 +906,20 @@ def post(self, request, *args, **kwargs):
...
+                # checking for users with pending/processing project removal requests.
+                username = user_form_data.get('username')
+                pending_status = ProjectUserRemovalRequestStatusChoice.objects.get(name='Pending')
+                processing_status = ProjectUserRemovalRequestStatusChoice.objects.get(name='Processing')
+
+                if ProjectUserRemovalRequest.objects.\
+                        filter(project_user__user__username=username,
+                               status__in=[pending_status, processing_status]).exists():
+
+                    message = f'Cannot add user {username} to project ' \
```
[the remark this answers is not in the document]
Comment (matthew-li) [reply coldfront/core/project/views.py:944]: @jofeinstein

**Comment by matthew-li** [comment]
Can you commit the migration files that were generated by `makemigrations`?

**Comment by matthew-li** [comment]
@jofeinstein Can you ensure that "Project Access on Cluster" is displayed as "Pending - Remove" on the home page if the user has a pending removal request?

(Their association technically still exists on the cluster at this point, but they can't submit jobs because the REST API checks that the `ProjectUser`'s status is "Active".)

**Comment by matthew-li** [comment]
I tried re-joining a project that I had left, and ran into `ProjectClusterAccessRequestRunner.validate_no_existing_cluster_access`.

The check exists so that we don't set someone who's already "Active" back to "Pending".

I think it makes sense to set this attribute's value to "Denied" when removing the user. (It might not exist for a PI who doesn't have a cluster account.)

**Comment by jofeinstein** [comment]
@matthew-li, In `ProjectRemovalRequestCompleteStatusView`, I set the `AllocationUserStatusChoice` to `Denied`. I also addressed the home page statuses along with several other small issues.

**Comment by matthew-li** [comment]
> @matthew-li, In `ProjectRemovalRequestCompleteStatusView`, I set the `AllocationUserStatusChoice` to `Denied`. I also addressed the home page statuses along with several other small issues.

@jofeinstein That's correct, but I'm referring to the `AllocationUserAttribute` with type `Cluster Account Status`, which denotes the status of their cluster access.

After a user joins a project, before automatically requesting cluster access, the method I mentioned above verifies that the user doesn't have already have "Active" cluster access on this project, since it would be set back to "Pending".

Setting the value to "Denied" when removing a user would prevent this check from getting triggered when this user attempts to re-join the project.

**Comment by jofeinstein** [comment]
Got it. I think I fixed it. Didn't run into any issues rejoining a project. I checked the values of `AllocationUserAttribute` with type `Cluster Account Status` after removing a user and after rejoining. Also checked to make sure all the tests still passed.

File: coldfront/core/project/views.py [hunk 10]
```diff
@@ -4162,13 +4162,15 @@ def form_valid(self, form):
                 allocation_obj = Allocation.objects.get(project=project_obj)
```
Comment (matthew-li) [inline coldfront/core/project/views.py:4162]: I'd move the `try`-`except` into the later `if status == 'Complete'` block, after the `ProjectUser`'s status gets updated.

So all the updates to the request happen in the same place, and we only update the allocation-related objects if the `ProjectUser` successfully got updated.

File: coldfront/core/project/views.py [hunk 11]
```diff
@@ -4162,13 +4162,15 @@ def form_valid(self, form):
...
                 allocation_user = \
                     allocation_obj.allocationuser_set.get(user=removed_user.user)
-                allocation_denied_status, _ = \
```
Comment (matthew-li) [inline coldfront/core/project/views.py]: I hadn't noticed that there is actually an `AllocationUserStatusChoice` called "Removed"; it seems appropriate to set that here as well.

That would also simplify this [block](https://github.com/ucb-rit/coldfront/pull/288/files#diff-5704bb865d9a2751bf4d43c1f88d4ce8d2bd8c175513f49a60d682e5bc33b67aR214); you could just filter for `AllocationUsers` with that status.

**Review by matthew-li — CHANGES_REQUESTED** [review]

**Comment by jofeinstein** [comment]
Made the above changes

File: coldfront/core/allocation/views.py [hunk 12]
```diff
@@ -151,6 +150,7 @@ def get_context_data(self, **kwargs):
                     except AttributeError:
                         pass
                 allocation_user_su_usages[username] = usage
+                print(username, usage)
```
Comment (matthew-li) [inline coldfront/core/allocation/views.py:153]: Some `print`s in this file.

(Did you configure these to show up somewhere?)

Comment (jofeinstein) [reply coldfront/core/allocation/views.py]: Yea, they get written to /etc/httpd/logs/cf_brc.error.log

**Comment by matthew-li** [comment]
Like in PR #302, emails should only get sent to PIs who have enabled notifications.

File: coldfront/core/project/tests/test_views/test_project_removal_request_views.py [hunk 13]
```diff
@@ -0,0 +1,650 @@
+from django.test import TestCase
```
Comment (matthew-li) [inline coldfront/core/project/tests/test_views/test_project_removal_request_views.py:1]: Can you test that the "Cluster Account Status" gets set to "Denied" and that the `AllocationUserStatusChoice` gets set to "Removed"?

This'll involve creating those allocation-related objects, which you may not have done before. See [here](https://github.com/ucb-rit/coldfront/pull/307/files#diff-7a5e4b8bc3fd0a15ae3e3a16913e7747ca7365a3999a83da77a7ab4b6336da4cR17) for an example.

(Long-term, I should add some test methods to make setting up these objects easier.)

**Review by matthew-li — APPROVED** [review]
Thanks for getting this done. I added a few more tweaks, but it looks good to my eye now.

We can get more feedback once this is deployed to staging.

**Pull request merged**